docs/developer-docs/build/backend part1 (#31)
Japan-DfinityInfoHub/portal: docs/developer-docs/build/backend/access-control.md

# Identity によるアクセスコントロールの追加

Dapps では、ユーザーごとに異なる操作を制御するために、ロールベースのパーミッションが必要になることがあります。

> このチュートリアルでは、ユーザー Identity の作成と切り替えを説明するために、異なるロールに割り当てられたユーザーに対して異なる greeding を表示するシンプルな Dapp を作成します。

Reviewer: greetingは訳しても違和感ないと思いました
```suggestion
このチュートリアルでは、ユーザー Identity の作成と切り替えを説明するために、異なるロールに割り当てられたユーザーに対して異なる挨拶文を表示するシンプルな Dapp を作成します。
```

この例では、`owner`、`admin`、`authorized` の3つのロールが指定されています。

> -   `admin` ロールが割り当てられているユーザーには、`You have a role with administrative privileges` と表示された greeding 文が表示されます。

Reviewer: ```suggestion
-   `admin` ロールが割り当てられているユーザーには、`You have a role with administrative privileges` という挨拶文が表示されます。
```

> -   `Authorized`ロールが割り当てられているユーザーには、`Would you like to play a game?` と表示された greeding 文が表示されます。

Reviewer: ```suggestion
-   `Authorized`ロールが割り当てられているユーザーには、`Would you like to play a game?` という挨拶文が表示されます。
```

> -   これらのロールが割り当てられていないユーザーには、`Nice to meet you!` と表示された greeding 文が表示されます。

Reviewer: ```suggestion
-   これらのロールが割り当てられていないユーザーには、`Nice to meet you!` という挨拶文が表示されます。
```

さらに、Canister を初期化したユーザー Identity にのみ `owner` ロールが割り当てられ、 `owner` と `admin` ロールのみが他のユーザーにロールを割り当てることができます。

> ハイレベルでは、各ユーザーは公開鍵と秘密鍵のペアを持ちます。ユーザーがアクセスする Canister ID に紐付いた公開鍵がセキュリティ Principal となり、これをメッセージ呼び出し元として使用して、Internet Computer ブロックチェーン上で動作する Canister への関数コールを認証することができます。次の図は、ユーザー Identity がメッセージ呼び出し元を認証する方法を簡略化して示しています。

Reviewer: 以下はどうでしょうか
```suggestion
大まかにいうと、各ユーザーは公開鍵と秘密鍵のペアを持ちます。ユーザーがアクセスする Canister ID に紐付いた公開鍵がセキュリティ Principal となり、これをメッセージ呼び出し元として使用して、Internet Computer ブロックチェーン上で動作する Canister への関数コールを認証することができます。次の図は、ユーザー Identity がメッセージ呼び出し元を認証する方法を簡略化して示しています。
```

![principal identities](../_attachments/principal-identities.svg)

## 始める前に

チュートリアルを始める前に、以下を確認してください：

-   [ダウンロードとインストール](../../quickstart/local-quickstart#download-and-install) で説明されているように、SDK パッケージをダウンロードしインストールする。

-   デフォルトのユーザー Identity が作成されるようなコマンドを少なくとも 1 つ実行する。デフォルトのユーザー Identity は、すべてのプロジェクトのために `$HOME/.config/dfx/identity/` ディレクトリにグローバルに保存されます。

-   Visual Studio Code を IDE として使用している場合、[言語エディタプラグインのインストール](../../quickstart/local-quickstart#install-vscode) にあるように、Motoko 用の Visual Studio Code プラグインをインストールする。

-   コンピュータ上で動作しているローカル Canister 実行環境プロセスを停止する。

## 新しいプロジェクトを作成する

アクセスコントロールとユーザー Identity の切り替えをテストするために、新しいプロジェクトディレクトリを作成します：

1.  ローカルコンピューターでターミナルシェルを開きます（まだ開いていない場合）。

> 2.  Internet Computer のブロックチェーン・プロジェクトに使用しているフォルダがある場合は移動します。

Reviewer: 少し読みにくさを感じました。以下はどうでしょうか。
```suggestion
2.  Internet Computer ブロックチェーンのプロジェクト用に使用しているフォルダがある場合にはそちらに移動します。
```

3.  以下のコマンドを実行し、新しいプロジェクトを作成します：

        dfx new access_hello

4.  次のコマンドを実行して、プロジェクトディレクトリに移動します：

        cd access_hello

## デフォルトの Dapp を変更する

このチュートリアルでは、テンプレートのソースコード・ファイルを、ロールの割り当てと取得のための機能を持つ Dapp に置き換えます。

デフォルトの Dapp を変更するには、次のようにします：

1.  `src/access_hello/main.mo` ファイルをテキストエディタで開き、既存のコンテンツを削除します。

2.  [このコード](../_attachments/access-control-main.mo)をコピーして、ファイルに貼り付けます。

    この Dapp の主要な要素をいくつか見てみましょう：

    -   これまでのチュートリアルで見てきた `greet` 関数のバリエーションであることに気がつくかもしれません。

>         しかし、このアプリでは、`greet` 関数がメッセージの呼び出し元を利用して、適用すべきパーミッションを決定し、呼び出し元に関連するパーミッションに基づいて、どの greeding を表示すべきかを決定します。

Reviewer: ```suggestion
        しかし、このアプリでは、`greet` 関数がメッセージの呼び出し元を利用して、適用すべきパーミッションを決定し、呼び出し元に関連するパーミッションに基づいて、どの挨拶文を表示すべきかを決定します。
```

## Dapp の登録、ビルド、デプロイ

開発環境で動作しているローカル Canister 実行環境に接続したら、[`dfx deploy`](../../../references/cli-reference/dfx-deploy) コマンドを実行して、一度にアプリの登録、ビルド、およびデプロイを行うことができます。また、[`dfx canister create`](../../../references/cli-reference/dfx-canister#dfx_canister_create)、 [`dfx build`](../../../references/cli-reference/dfx-build) と [`dfx canister install`](../../../references/cli-reference/dfx-canister#dfx_canister_install) コマンドでこれらの手順を個別に行うことが可能です。

Dapp をローカルにデプロイするには：

1.  必要に応じて、まだプロジェクトのルートディレクトリにいることを確認します。

2.  以下のコマンドを実行して、`access_hello` バックエンド Dapp を登録、ビルド、デプロイします：

        dfx deploy access_hello

        Creating a wallet canister on the local network.
        The wallet canister on the "local" network for user "default" is "rwlgt-iiaaa-aaaaa-aaaaa-cai"
        Deploying: access_hello
        Creating canisters...
        Creating canister "access_hello"...
        "access_hello" canister created with canister id: "rrkah-fqaaa-aaaaa-aaaaq-cai"
        Building canisters...
        Installing canisters...
        Installing code for canister access_hello, with canister_id rrkah-fqaaa-aaaaa-aaaaq-cai
        Deployed canisters.

## 現在の Identity コンテキストを確認する

> 追加の Identity を作成する前に、`default` Identity に関連する Principal 識別子と `default` Identity の Cycle ウォレットを確認しましょう。Internet Computer ブロックチェーンでは、 Principal はユーザー、Canister、ノード、サブネットの内部代表です。 Principal のテキスト表現は、 Principal データタイプで作業しているときに表示される外部識別子です。

Reviewer: 内部代表が違和感ありました。以下はどうでしょうか
```suggestion
追加の Identity を作成する前に、`default` Identity に関連する Principal 識別子と `default` Identity の Cycle ウォレットを確認しましょう。Internet Computer ブロックチェーンでは、 Principal はユーザー、Canister、ノード、サブネットを識別するための内部的な代表値です。 Principal のテキスト表現は、 Principal データタイプで作業しているときに表示される外部識別子です。
```
